crates/block-producer/src/batch_builder/batch.rs

```diff
@@ -44,38 +47,50 @@ impl TransactionBatch {
     pub fn new(txs: Vec<ProvenTransaction>) -> Result<Self, BuildBatchError> {
         let id = Self::compute_id(&txs);
 
-        // TODO: we need to handle a possibility that a batch contains multiple transactions against
-        //       the same account (e.g., transaction `x` takes account from state `A` to `B` and
-        //       transaction `y` takes account from state `B` to `C`). These will need to be merged
-        //       into a single "update" `A` to `C`.
-        let updated_accounts =
-            txs.iter().map(|tx| (tx.id(), tx.account_update().clone())).collect();
+        let mut updated_accounts = vec![];
+        let mut produced_nullifiers = vec![];
+        let mut unauthenticated_input_notes = BTreeSet::new();
+        for tx in &txs {
+            // TODO: we need to handle a possibility that a batch contains multiple transactions against
+            //       the same account (e.g., transaction `x` takes account from state `A` to `B` and
+            //       transaction `y` takes account from state `B` to `C`). These will need to be merged
+            //       into a single "update" `A` to `C`.
+            updated_accounts.push((tx.id(), tx.account_update().clone()));
+
+            for note in tx.input_notes() {
+                produced_nullifiers.push(note.nullifier());
+                if let Some(note_id) = note.note_id() {
+                    unauthenticated_input_notes.insert(note_id);
```

Comment:
@bobbinth, according to my current solution, unauthenticated note `x` would be removed from the set on the first consuming and the second attempt to consume it would raise an error. But let's also forbid producing note with the same ID twice, it might help with debugging of such cases.